Schema changes for persisted metadata

TommyBrosman opened the pull request:
(TODO: description)

TommyBrosman committed:
Initial change for the persisted metadata feature.
- Introduces a new Stored Schema format that has a persisted field. formatV2.ts is dead code for now and not loaded anywhere.
- Added the persisted metadata field to the View Schema (see the change to schemaTypes.ts). It is not wired up yet.

TommyBrosman committed:
Created a new Format for the schema codec. This one includes the persisted value.

TommyBrosman committed:
- Introduced an enum for the schema format version. It is not practical to use the Shared Tree write format as there may be multiple write formats for a single schema format. As a consequence of this change, all callers to makeSchemaCodec now specify the SchemaFormatVersion.
- Changed makeSchemaChangeCodecs to include a version for write format 4. Not wired up (or tested) yet.
- Forked encodeRepo and created a facade to call the versioned functions.

TommyBrosman commented on packages/dds/tree/src/simple-tree/schemaTypes.ts:
```diff
@@ -1002,4 +1002,9 @@ export interface NodeSchemaMetadata<out TCustomMetadata = unknown> {
+	/**
+	 * Metadata defined at schema creation time.
+	 */
+	readonly persistedData?: string | undefined;
```
TODO: Replace with a more specific type (e.g., `Map<string, string>`).

TommyBrosman commented on packages/dds/tree/src/feature-libraries/schema-edits/schemaChangeFormat.ts:
```diff
@@ -5,7 +5,7 @@
 import { type Static, Type } from "@sinclair/typebox";
 
-import { Format } from "../schema-index/index.js";
+import { FormatV1 as Format } from "../schema-index/index.js";
```
Should this support multiple Formats? The version checked in changes the contents of the Format and not necessarily the Format itself.

CraigMacomber replied:
If doing schema edits doesn't support both formats, then there is no way to use the new format to get the desired metadata into a document, so yes, this code for editing the schema of a document needs to support both formats.

TommyBrosman committed:
Minor: cleaned up switch statements so they use the `unreachableCase` method.

TommyBrosman committed:
- Updated a few tests to cover both formats. There are still many incorrect snapshots, and I need to make a pass over all of the tests that touch schema formats.
- Removed an extraneous codec I introduced in schemaChangeCodec. The schema versioning is sufficient.

CraigMacomber commented on packages/dds/tree/api-report/tree.alpha.api.md:
```diff
@@ -812,6 +813,14 @@ export interface SchemaFactoryObjectOptions<TCustomMetadata = unknown> extends N
     allowUnknownOptionalFields?: boolean;
 }
 
+// @alpha
```
We do not want to expose details of our persisted formats in the app facing API surface. Due to having to do compat for all past versions, details like this will inherently grow in complicity over time and we do not want to commit to making the app facing APIs reflect that complexity.

User/app facing APIs like extractPersistedSchema should only care about what versions of clients something works with, not which format is used and should thus use FluidClientVersioninstead.

TommyBrosman replied:
I need to specify which version of the schema I am using when I create codecs/etc. Is your suggestion to pass in the FluidClientVersion there and not expose this type?

CraigMacomber commented on packages/dds/tree/api-report/tree.alpha.api.md:
```diff
@@ -130,7 +130,7 @@ export function evaluateLazySchema<T extends TreeNodeSchema>(value: LazyItem<T>)
 
 // @alpha
-export function extractPersistedSchema(schema: ImplicitFieldSchema): JsonCompatible;
+export function extractPersistedSchema(schema: ImplicitFieldSchema, version: SchemaFormatVersion): JsonCompatible;
```
Like exportCompressed, this should probably take in an options object so we can provide more settings later without adding lots of arguments, and take in the version as FluidClientVersion, which is already exposed as alpha  `options: {
        oldestCompatibleClient: FluidClientVersion;
    }`
    
    Look at /examples/apps/tree-cli-app/src/utils.ts (who's build you broke with this API change) for an example usage. It already uses exportCompressed with FluidClientVersion and calls extractPersistedSchema: passing in the same FluidClientVersion there would be nicer and clean compared to taking in a different version specific for schema stuff.

CraigMacomber commented on packages/dds/tree/api-report/tree.alpha.api.md:
```diff
@@ -812,6 +813,14 @@ export interface SchemaFactoryObjectOptions<TCustomMetadata = unknown> extends N
 
+// @alpha
+export enum SchemaFormatVersion {
+    // (undocumented)
```
these members should probably have docs. V1 likley documenting that its supported in all clients which support shared tree (2.0 +)

and V2 (which should maybe be renamed to VUnstable) documenting that its format is not yet stable and thus should not be used for data which needs to be supported by future versions. When this code is actually all finished end to end, then it can be updated to V2, and have the version it will get released in noted.

CraigMacomber commented on packages/dds/tree/src/core/schema-stored/formatV2.ts:
```diff
@@ -0,0 +1,104 @@
+ *
+ * Stable identifier, used when persisting data.
+ */
+export type FieldKey = Brand<string, "tree.FieldKey">;
```
There is a lot of duplication for things that are the same in v1 and v2. At least while v2 is under development, I think it would make sense for v2 to just import things from v1 so the changes are more clear, and we don't duplicate the docs and such.

Someday after V2 is finalized we might want to refactor it such that v1 imports from v2 so that the current version is the cleanest and the legacy stuff references it, but I think the reverse (as proposed above) is best for now.

TommyBrosman replied:
I like that approach. Forking it for now makes the separation obvious, but I can update it to import from V1 later in this PR.

CraigMacomber commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -120,6 +137,11 @@ export interface TreeFieldStoredSchema {
 	readonly types: TreeTypeSet;
+
+	/**
+	 * Persisted metadata. Part of the v2 stored schema format.
```
```suggestion
	 * Portion of the metadata which can be persisted.
	 * @remarks
	 * Discarded when encoding to {@link SchemaFormatVersion.V1}.
```

CraigMacomber commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -155,15 +178,16 @@ export interface ErasedTreeNodeSchemaDataFormat
 	return data as unknown as ErasedTreeNodeSchemaDataFormat;
 }
 
+// TODO: Should this always return a V1 Format?
```
ErasedTreeNodeSchemaDataFormat only exists to type erase the schema format returned by TreeNodeStoredSchema.encode.

Since we removed the package export for TreeNodeStoredSchema, its reason for existance (and thus this function existing) is gone.

TreeNodeStoredSchema.encode should be made to take in SchemaFormatVersion, return TreeNodeSchemaDataFormatV1 | TreeNodeSchemaDataFormatV2 and ErasedTreeNodeSchemaDataFormat and the related functions to cast to and from it should be removed. Ideally do this cleanup in a separate PR (using just TreeNodeSchemaDataFormat since v2 won't exist yet).

CraigMacomber commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -203,7 +227,7 @@ export class ObjectNodeStoredSchema extends TreeNodeStoredSchema {
 	}
 
 	public override encode(): ErasedTreeNodeSchemaDataFormat {
```
I'm not sure how you intended to support encoding into the new format if this encode method doesn't take in the client version and return a union of all the allowed formats.

TommyBrosman replied:
Why the client version in this case? I'd think that the ST format or schema format would be a better fit. The codecs I've looked at so far all use the ST format version for dispatch.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/schema-index/codec.ts:
```diff
@@ -37,9 +73,10 @@ export function encodeRepo(repo: TreeStoredSchema): Format {
 		});
```
I can't comment on it (thanks git hub) but the line before this is using the schema.encode(), which is the v1 format. You only used the v2 format for the root field: you need to use it for everything

TommyBrosman replied:
Nice catch, I hadn't looked at the implementations of `TreeNodeStoredSchema.encode` yet. I'll have to update the implementations to produce a different format depending on write version. Fortunately it is known here. I'll probably just change the API to take a Shared Tree write format.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/schema-index/codec.ts:
```diff
@@ -57,9 +94,32 @@ function decode(f: Format): TreeStoredSchema {
-		encode: (data: TreeStoredSchema) => encodeRepo(data),
-		decode: (data: Format) => decode(data),
-	});
+export function makeSchemaCodec(
```
I don't think you want to use makeVersionedValidatedCodec. When encoding schema in v2, you don't want a codec which will error when trying to read v1 data (which is what this will give you)

I think you want to use makeVersionDispatchingCodec. I haven't used it, but it looks like the thing you want for making a coded which supports multiple formats.

CraigMacomber commented on packages/dds/tree/src/test/snapshots/files/SchemaIndexFormatFormat_1.json:
```diff
@@ -0,0 +1,171 @@
+  "type": "object",
+  "properties": {
+    "version": {
+      "const": 2,
```
this file name is _1 and version 2, and version 1 is in filer _0. Thats confusing.

TommyBrosman replied:
I'm standardizing everything to have `FormatV1`/`FormatV2` suffixes.

CraigMacomber commented:
It might be nice to first do PRs for:
1. Removing ErasedTreeNodeSchemaDataFormat  (as suggested in a comment on the relevant code)
2. Adding in support for more versions (ex: renaming tests/snapshots, adopt makeVersionDispatchingCodec, pass in FluidClientVersion where needed etc. Note that if you do thee before actually adding a format v2, then it can be done in smaller parts (ex: one PR adopting makeVersionDispatchingCodec, one renaming the snapshot tests, one passing in the FluidClientVersion. That should make review a lot easier.

TommyBrosman commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -10,13 +10,30 @@ import { type MakeNominal, brand, invertMap } from "../../util/index.js";
+export enum SchemaFormatVersion {
+	V1 = 1,
+	V2 = 2,
+}
```
This might go away. We probably need to specify the format version as a parameter somewhere, but it doesn't necessarily need to be an enum. It's weird that we would have a separate version number for the schema in the codecs.

TommyBrosman replied:
We could also just specify this version at the top and translate it into the codec's schema version.

TommyBrosman committed:
Rewrote the schema codec and plumbed it through the SchemaSummarizer.

TommyBrosman committed:
Fixed issue with version numbers. The versions specified when constructing the codec should correspond to the schema format version and not the top-level Shared Tree write format.

TommyBrosman committed:
- Removed SchemaFormatVersion enum and made versioning match existing Shared Tree conventions.
- Updated SharedTreeKernel to pass the correct version instead of the top-level Shared Tree format.

TommyBrosman committed:
Made the schemaChangeCodec handle both formats.

TommyBrosman committed:
Cleaned up and documented the schema codec.

TommyBrosman committed:
- Updated test trees so that it generates trees in both schema formats. Updated snapshots for schema-files.
- Plumbed a write version parameter through for `comparePersistedSchema`.

TommyBrosman committed:
Minor refactor.

TommyBrosman committed:
Fixed most of the non-fuzz snapshot tests.

TommyBrosman committed:
Added generated tests in schema.spec.ts for supported schema formats.

TommyBrosman committed:
Update packages/dds/tree/src/core/schema-stored/schema.ts

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

TommyBrosman commented on packages/dds/tree/src/feature-libraries/schema-index/schemaSummarizer.ts:
```diff
@@ -116,6 +114,6 @@ export class SchemaSummarizer implements Summarizable {
-export function encodeTreeSchema(schema: TreeStoredSchema): Format {
-	return encodeRepo(schema);
+export function encodeTreeSchema(schema: TreeStoredSchema, version: number): JsonCompatible {
+	return encodeRepo(schema, version as 1 | 2);
```
I'm considering removing the version type that explicitly calls out the supported versions. It seems like it will be difficult to maintain in the long-term given the number of methods that contain this parameter. Thoughts/opinions?

TommyBrosman commented on packages/dds/tree/src/shared-tree/independentView.ts:
```diff
@@ -92,7 +92,7 @@ export function independentInitializedView<const TSchema extends ImplicitFieldSc
 
 	const fieldBatchCodec = makeFieldBatchCodec(options, 1);
-	const schemaCodec = makeSchemaCodec(options);
+	const schemaCodec = makeSchemaCodec(options, 1);
```
Should independentView handle both formats?

TommyBrosman commented on packages/dds/tree/src/core/tree/persistedTreeTextFormat.ts:
```diff
@@ -5,7 +5,7 @@
 import { type Static, type TSchema, Type } from "@sinclair/typebox";
 
-import { schemaFormat } from "../schema-stored/index.js";
+import { schemaFormatV1 as schemaFormat } from "../schema-stored/index.js";
```
Todo: Make it more obvious that this code doesn't care about schema format versions, just TreeNodeSchemaIdentifierSchema specifically.

TommyBrosman commented:
See !24590 and !24479.
